server/commands/write.go

```diff
@@ -71,32 +71,41 @@ func (c *WriteCommand) validateTuplesets(ctx context.Context, req *openfgapb.Wri
 		if err != nil {
 			return err
 		}
+
+		typesys = typesystem.New(authModel)
 	}
 
 	for _, tk := range writes {
-		tupleUserset, err := validation.ValidateTuple(ctx, c.datastore, store, modelID, tk)
+		err := validation.ValidateTuple(typesys, tk)
 		if err != nil {
 			return serverErrors.HandleTupleValidateError(err)
 		}
 
-		// Validate that we are not trying to write to an indirect-only relationship
-		if !typesystem.RewriteContainsSelf(tupleUserset) {
-			return serverErrors.HandleTupleValidateError(&tupleUtils.IndirectWriteError{Reason: IndirectWriteErrorReason, TupleKey: tk})
-		}
+		objectType, _ := tupleUtils.SplitObject(tk.GetObject())
 
-		if err := c.validateNoUsersetForRelationReferencedInTupleset(authModel, tk); err != nil {
-			return err
+		relation, err := typesys.GetRelation(objectType, tk.GetRelation())
+		if err != nil {
+			if errors.Is(err, typesystem.ErrObjectTypeUndefined) {
+				return serverErrors.TypeNotFound(objectType)
+			}
+
+			if errors.Is(err, typesystem.ErrRelationUndefined) {
+				return serverErrors.RelationNotFound(tk.GetRelation(), objectType, tk)
+			}
+
+			return serverErrors.HandleError("", err)
 		}
 
-		if err := c.validateTypesForTuple(authModel, tk); err != nil {
-			return err
+		// Validate that we are not trying to write to an indirect-only relationship
+		if !typesystem.RewriteContainsSelf(relation.GetRewrite()) {
+			return serverErrors.HandleTupleValidateError(&tupleUtils.IndirectWriteError{Reason: IndirectWriteErrorReason, TupleKey: tk})
 		}
 	}
 
 	for _, tk := range deletes {
 		// For delete, we only need to ensure it is well form but no need to validate whether relation exists
```

Comment:
While you are here, do you mind fixing this comment, or just deleting it?

Reply:
Yep! Removed. Seemed redundant, I agree.